File: Έγγραφα SEM4105/Λύσεις ασκήσεων/08. Όγδοο εργαστήριο/2/stefbol.cpp
#include <vector>
#include <iostream>
#include <numeric>
#include <fstream>
#include <functional>
#include <cmath>
#include <algorithm>
#include <string>


using vec = std::vector<double>;

double mylog(double x)
{
    return std::log(x);
}

bool allpositive(vec const & x)
{
    auto gt0 = [] (auto & x) -> bool { return x > 0; };
	
    return std::all_of(x.cbegin(), x.cend(), gt0);
}


//     Read data file
void read_data(std::string const & fname, vec & x, vec & y)
{
    std::ifstream in{fname};

    int n;
    in >> n;
    
    x.reserve(n);
    y.reserve(n);

    for (int i{0}; i < n; ++i) {
	double a, b;
	in >> a >> b;
	x.push_back(a);
	y.push_back(b);
    }
}


//     Y = A1 * X + a0
void lsqlin(vec const & x, vec const & y,
	    double & a1, double & a0, double & r2)
{
    auto const n = x.size();

    auto const sx = std::accumulate(x.cbegin(), x.cend(), 0.0);
    auto const sy = std::accumulate(y.cbegin(), y.cend(), 0.0);
    auto const sx2 = std::inner_product(x.cbegin(), x.cend(), x.cbegin(), 0.0);
    auto const sy2 = std::inner_product(y.cbegin(), y.cend(), y.cbegin(), 0.0);
    auto const sxy = std::inner_product(x.cbegin(), x.cend(), y.cbegin(), 0.0);

    auto const d = n * sx2 - sx * sx;

    a1 = (n * sxy - sx * sy) / d; 
    a0 = (sy - a1 * sx) / n;

    r2 = std::pow(n * sxy - sx * sy,2) / d / (n * sy2 - sy*sy);
}




//     Y = a0*X^A1 => LN(Y) = LN(a0) + A1 * LN(X)
void lsqpow(vec const & x, vec const & y,
	    double & a1, double & a0, double & r2)
{
    if (!allpositive(x) || !allpositive(y)) {
	std::cerr << "Data not appropriate\n";
	return; 
    }

    auto const n = x.size();
    vec zx(n);
    vec zy(n);
    double za0;

    std::transform(x.cbegin(), x.cend(), zx.begin(), mylog);
    std::transform(y.cbegin(), y.cend(), zy.begin(), mylog);    

    lsqlin(zx, zy, a1, za0, r2);
  
    a0 = std::exp(za0);
}


int main()
{
    vec x, y;
    read_data("stefbol.dat", x,y);

    double a1, a0, r2;
    lsqpow(x, y, a1, a0, r2);
    
    double constexpr embado{0.05e-4};

    std::cout << u8"Σταθερά Stefan-Boltzmann: "
	      << a0 / embado << " J/K^4/m^2/s\n";
    std::cout << u8"Εκθέτης: " << a1 <<'\n';
    std::cout << "r^2 = " << r2 << '\n';
}

File: Έγγραφα SEM4105/Λύσεις ασκήσεων/08. Όγδοο εργαστήριο/2/stefbol.f
      PROGRAM STEFBO
      INTEGER N
      DOUBLE PRECISION A1, A0, X(1000), Y(1000), R2

      DOUBLE PRECISION EMBADO

      CALL READAT("stefbol.dat", N, X, Y)

C     Y = A0 X^A1
      CALL LSQPOW(N, X, Y, A1, A0, R2)

      EMBADO = 0.05D-4

      PRINT *, "Σταθερά Stefan-Boltzmann: ", 
     &        A0 / EMBADO, "J/K^4/m^2/s"
      PRINT *, "Εκθέτης: ",  A1
      PRINT * , "r^2 = ", R2
      END


C     Read data file
      SUBROUTINE READAT(FNAME, N, X, Y)
      CHARACTER(*) FNAME
      INTEGER N
      DOUBLE PRECISION X(*), Y(*)

      INTEGER UNITNO, I
      PARAMETER (UNITNO = 54)

      OPEN (UNIT = UNITNO, FILE = FNAME)
      READ(UNITNO, *) N

      DO 10 I=1,N
        READ (UNITNO, *) X(I), Y(I)
 10   CONTINUE

      CLOSE (UNITNO)
      END


C     Y = A1 * X + A0
      SUBROUTINE LSQLIN(N, X, Y, A1, A0, R2)
      INTEGER N
      DOUBLE PRECISION X(*), Y(*), A1, A0, R2

      DOUBLE PRECISION SX, SY, SX2, SY2, SXY, D
      INTEGER I

      SX  = 0.0D0
      SY  = 0.0D0
      SX2 = 0.0D0
      SY2 = 0.0D0
      SXY = 0.0D0

      DO 10 I = 1, N
        SX  = SX  + X(I)
        SY  = SY  + Y(I)
        SX2 = SX2 + X(I) * X(I)
        SY2 = SY2 + Y(I) * Y(I)
        SXY = SXY + X(I) * Y(I)
 10   CONTINUE

      D = N * SX2 - SX * SX

      A1 = (N * SXY - SX * SY) / D 
      A0 = (SX2 * SY - SX * SXY) / D 

      R2 = (N * SXY - SX * SY)**2 / (N * SX2 - SX**2)/(N * SY2 - SY**2)
      END


C     Y = A0 X^A1 => LN(Y) = LN(A0) + A1 * LN(X)
      SUBROUTINE LSQPOW(N, X, Y, A1, A0, R2)
      INTEGER N     
      DOUBLE PRECISION X(*), Y(*), A1, A0, R2

      DOUBLE PRECISION ZX(N), ZY(N), ZA0
      INTEGER I

      DO 10 I = 1, N
        IF ( (X(I) .LE. 0.D0) .OR. (Y(I) .LE. 0.D0) ) THEN
           PRINT *, "Data not appropriate"
           RETURN
        ENDIF

        ZX(I) = LOG(X(I))
        ZY(I) = LOG(Y(I))
 10   CONTINUE
      
      CALL LSQLIN(N, ZX, ZY, A1, ZA0, R2)

      A0 = EXP(ZA0)

      END

